### ISSUE TITLE
flutter license check comes too late in the install process, doctor doesn't warn about it

### ISSUE THREAD

**Issue by sethladd** [flutter/flutter#13584]
I'm installing flutter on a brand-new mac. I ran into a surprising situation, which I think we can make much better for new users.

tldr: flutter doctor says everything is fine, however things aren't fine: I didn't accept my licenses.

Proposal: have doctor check if I accepted my licenses, prompt me to do it _before_ I run my first app.

(apparently we know if the user accepted the licenses, because we tell the user to run "flutter doctor --android-licenses)

Here's what I see before I run my first app:

```
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.2)
    • Android SDK at /Users/sethladd/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)
```

However, here's what I'm told when I try to run my first app:

```
sethladd-macbookpro:my_first_app sethladd$ flutter run
Launching lib/main.dart on Android SDK built for x86 in debug mode...
Initializing gradle...                                9.6s
Resolving dependencies...                                -

* Error running Gradle:
Unable to download needed Android SDK components, as the following licenses have not been accepted:
Android SDK Build-Tools 25.0.3, Android SDK Platform 25

To resolve this, please run the following command in a Terminal:
flutter doctor --android-licenses
\sethladd-macbookpro:my_first_app sethladd$ flutter doctor --android-licenses
Warning: File /Users/sethladd/.android/repositories.cfg could not be loaded.    
5 of 6 SDK package licenses not accepted. 100% Computing updates...             
Review licenses that have not been accepted (y/N)? y

```

Sad user, I thought I was good to go, because I had the checkmark, but roadbump!

I think our users would much rather see this warning the first time they run `flutter doctor`, at least when/while they run `flutter doctor` as they are setting up their Android environment, because that's when they are getting their dependencies setup.

The first time I run, assuming doctor gives me a checkmark, my app should _just run_.

Thanks for taking a look!

**Comment by sethladd** [comment]
cc @mit-mit @devoncarew

**Comment by mit-mit** [comment]
We don't actually know if there are any missing license accepts prior to doing a build; that message is shown after catching an exception in a failing build.

I guess we could do a build of a dummy build as part of doctor, but that would add ~30 seconds to doctor, which I'm not sure is worth it.

I think I'd rather spend our time on preventing this (not having the needed sdk components installed) in the first place.

**Comment by mit-mit** [comment]
Actually, even further: currently we use whatever buildtools version is listed in the config files in the android folder. These may differ from app to app, so generally doctor can't guarantee that any project will build.

I think the important part here is to check that new projects created by `flutter create` will build.

**Comment by mit-mit** [comment]
cc @mravn-google I think this is the issue we discussed earlier today?

**Comment by jcollins-g** [comment]
First impression is that sdkmanager's --license command could have its input redirected from /dev/null, and it will then display the number of licenses not yet accepted.  This only takes a couple of seconds to run, so fits with doctor, and could be used to display a "missing" X on Android in doctor.

**Comment by mit-mit** [comment]
Sounds like a good solution!

**Comment by jcollins-g** [comment]
Implemented, but the couple of seconds that sdkmanager's license command takes puts a noticeable delay in flutter doctor's output.  Seeing if there's any way to mitigate that.

**Comment by mit-mit** [comment]
I think a delay is ok. If it feels long maybe put a progress spinner on it?

**Comment by jcollins-g** [comment]
Mitigations were unsuccessful; the rest of doctor is so fast that making things asynchronous doesn't help noticeably.

As for whether it "feels long", it's marginal, and it looks like sdkmanager might be doing some network accesses before getting the licenses printed so it could vary for different users.  So, adding a spinner.

**Comment by jcollins-g** [comment]
I have a version of doctor that seems to work for me and passes existing tests; going to write some more and clean things up.  Here's what the initial version looks like:

With no licenses installed, -v to show the info text:
```
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /usr/local/google/home/jcollins/Android/Sdk
    • Android NDK at /usr/local/google/home/jcollins/Android/Sdk/ndk-bundle
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /opt/android-studio-3.0/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)
    • All Android licenses accepted.
```

With some fraction of them installed (I've had android builds work fine with at least some licenses accepted), without -v:
```
[!] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses
```

With no licenses, no -v:
```
[!] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    ✗ Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses
```

**Comment by jcollins-g** [comment]
Reopened as the fix proved problematic in its first incarnation, was reverted in #14727.

Doctor on Mac OS X pops up a JDK prompt when no JDK is installed:  
 https://github.com/flutter/flutter/pull/14535#issuecomment-365891126
Doctor on Mac OS X sometimes causes a hang: https://github.com/flutter/flutter/issues/14725 
 even when Java is installed, symptom likely depends on whether the user is in a position to see the window. 
Spinner has extra character on the end of lines:  https://github.com/flutter/flutter/pull/14535#issuecomment-365792937

AnsiStatus.cancel() printing elapsed time and extra newline:  https://github.com/flutter/flutter/pull/14535#issuecomment-365796339

Don't know the cause of the hang yet, but the other issues are pretty clear.  Will find/fix them all and retry.

**Comment by mit-mit** [comment]
@jcollins-g about the java issue from https://github.com/flutter/flutter/pull/14535#issuecomment-365891126: I updated after your review, and now doctor works. Interestingly it reports that I do have java. So I'm wondering if we just somehow need to tell sdkmanager where to find java?

```
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/mit/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)
```

**Comment by jcollins-g** [comment]
I have fixes, pending testing, for all issues reported.   sdkmanager wasn't being set up with the right environment to find the right Java which is why you are being prompted.

**Comment by jcollins-g** [comment]
OK, so this is taking far longer than I had hoped.  Part of the reason is that the unittests currently don't mock out process execution in many places where they need to start doing it if we're going to fork and exec something in the middle of doctor (up until this point, doctor hasn't ever needed to do this).   Some of them depend on not mocking it out as written, while others depend on mocking it in specific ways, etc.

Determining exactly how to mock this out is going to require reading a lot of code, determining the best way to cover this in all cases, and rearranging things so that works, possibly with multiple PRs.
